Beautify code

andrew000 opened the pull request:
- Move project to `UV`; https://docs.astral.sh/uv
- Add static type checker `MyPy`
- Add linter and code formatter `ruff`
- Add `pre-commit`
- Reformat code using `ruff`
- Make code more secure (stop using `os.path`, `os.system`)
- Update workflow `python_simplified.yml`
- Update workflow `python_detailed.yml`
- Drop py3.8, because of EOL.

Drop workflow python version from 3.10 to 3.9 to be sure that the library will work on py3.9+. Also, I want to add `matrix` to test the library on 3.9, 3.10, 3.11, 3.12, 3.13, but I don't see anything deprecated in 3.9 that can't be in 3.13 in this code.

andrew000 commented on oqs/oqs.py:
```diff
@@ -247,102 +296,131 @@ def __init__(self, alg_name, secret_key=None):
-        return bytes(ciphertext), bytes(shared_secret) if rv == OQS_SUCCESS else 0
 
-    def decap_secret(self, ciphertext):
+        # TODO: What should it return?
+        #  1. tuple[bytes | int, bytes | int]
+        #  2. tuple[bytes, bytes | int]
+        #  3. tuple[bytes, bytes] | int
+        return (
+            bytes(cast(bytes, ciphertext)),
+            bytes(cast(bytes, shared_secret)) if rv == OQS_SUCCESS else 0,
+        )
```
What fuction should return?

vsoftco replied:
this was changed in the last version, returning only tuple(bytes, bytes) and throwing if not successful

Salamist commented on oqs/oqs.py:
```diff
@@ -1,29 +1,43 @@
 
-def oqs_python_version():
+
+def oqs_python_version() -> str | None:
```
Hello,
Using pipe operator as shorthand for Union will not work on Python 3.9 (This feature was introduced in Python3.10.

vsoftco commented:
I will merge this after 0.12 release (later this week)

andrew000 commented:
> I will merge this after 0.12 release (later this week)

After accepting #100 I will refactor this PR to apply changes from #100 .

vsoftco commented:
@andrew000 We've released 0.12. Updated the default examples to use ML-KEM/DSA. Please feel free to resolve the other conflicts then we can merge the PR.

andrew000 commented:
I will resolve conflicts later this week

vsoftco commented:
@andrew000 Thanks, I will review shortly and merge

vsoftco commented:
@andrew000 I think we need something like `logger.addHandler(logging.StreamHandler(sys.stdout))` in the examples, so we can filter out warnings from `oqs` (like the mismatch between oqs and oqs-python) with `python examples/kem.py 2>/dev/null`

vsoftco commented:
@andrew000 I've modified the examples. Let me know if this looks OK for you, and I'll merge the PR.

andrew000 commented:
> @andrew000 I've modified the examples (see the last commit). Let me know if this looks OK for you, and I'll merge the PR.

Tonight I will review the changes and fix some things if I find anything.

vsoftco commented:
@andrew000 I'm getting some Pyright type errors for `_fields_`, L270 and L467:

`"_fields_" overrides symbol of same name in class "Structure"
     │     Variable is mutable so its type is invariant
     │       Override type "list[tuple[str, Any]]" is not the same as base type "Sequence[tuple[str, type[_SimpleCData[Any]] | type[_Pointer[Any]] | type[CFuncPtr] | type[Union] | type[Structure] | type[Array[Any]]] | tuple[str, type[_SimpleCData[Any]] | type[_Pointer[Any]] | type[CFuncPtr] | type[Union] | type[Structure] | type[Array[Any]], int]]" Pyright (reportIncompatibleVariableOverride) [270, 5]`

I guess it's the static type checker being unhappy, but I'm not too familiar with types in Python. Any clues?

andrew000 commented:
> @andrew000 I'm getting some Pyright type errors for `_fields_`, L270 and L467 in `oqs.py`:
> 
> `"_fields_" overrides symbol of same name in class "Structure" │   Variable is mutable so its type is invariant │     Override type "list[tuple[str, Any]]" is not the same as base type "Sequence[tuple[str, type[_SimpleCData[Any]] | type[_Pointer[Any]] | type[CFuncPtr] | type[Union] | type[Structure] | type[Array[Any]]] | tuple[str, type[_SimpleCData[Any]] | type[_Pointer[Any]] | type[CFuncPtr] | type[Union] | type[Structure] | type[Array[Any]], int]]" Pyright (reportIncompatibleVariableOverride) [270, 5]`
> 
> I guess it's the static type checker being unhappy, but I'm not too familiar with types in Python. Any clues?

https://docs.python.org/3/library/ctypes.html#ctypes.Structure

> _Concrete structure and union types must be created by subclassing one of these types, and at least define a \_fields\_ class variable._

andrew000 commented:
> @andrew000 I've modified the examples (see the last commit). Let me know if this looks OK for you, and I'll merge the PR.

LGTM. 

I have updated `ruff` in last commit.

So I think PR is ready to be merged 👍

andrew000 commented:
Documentation does not provide any detailed type hints. 

`typeshed` declaring a type hint:
```python
class _StructUnionMeta(_CDataMeta):
    _fields_: Sequence[tuple[str, type[_CData]] | tuple[str, type[_CData], int]]
```

But `_CData` is not declared in python sources.

Finally we can try change `list` to `Sequence` to try to calm down static type checker
